src/iocost/iocost.c

```diff
@@ -0,0 +1,145 @@
+#include "verbs.h"
+
+static int help(void) {
+        _cleanup_free_ char *link = NULL;
+        int r;
+
+        r = terminal_urlify_man("systemd-iocost", "1", &link);
+        if (r < 0)
+                return log_oom();
+
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices\n"
+               "\nCommands:\n"
+               "  apply <path or node>       Apply the known solution for the device, if any, otherwise does nothing\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n"
+               "\nSee the %s for details.\n",
+               program_invocation_short_name,
+               link);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int apply_solution_for_path(const char *path) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *qos = NULL, *model = NULL;
+        const char *params;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        r = sd_device_get_property_value(device, "IOCOST_TUNE_MODEL", &params);
+        if (r < 0) {
+                log_info("Parameters for iocost model not available, doing nothing.");
+                return 0;
+        }
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), params) < 0)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, "IOCOST_TUNE_QOS", &params);
+        if (r < 0) {
+                log_info("Parameters for iocost qos not available, doing nothing.");
+                return 0;
+        }
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), params) < 0)
+                return log_oom();
```

Comment:
not that it matters much, but i'd do the two sd_device_get_property_value() calls first, so that you exit early if no data is available. And only then I'd start formatting them into strings to write to the sysfs attributes.

i mean, it doesn't matter performance-wise, this is not an inner loop. But it is confusing already processing loaded data before you know it actually is complete. At least to me it is confusing.

Reply:
The new version factors that out into a function that is used by both query and apply verbs, so this should be no longer an issue =)